hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// createStorageTypes returns a pipeline stage that creates dedicated storage types for each resource and nested object.
+// Storage versions are created for *all* API versions to allow users of older versions of the operator to easily
+// upgrade. This is of course a bit odd for the first release, but defining the approach from day one is useful.
+func createStorageTypes() PipelineStage {
+	return PipelineStage{
+		id:          "createStorage",
+		description: "Create storage versions of CRD types",
+		Action: func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+
+			storageTypes := make(astmodel.Types)
+			visitor := makeStorageTypesVisitor(types)
+			vc := makeStorageTypesVisitorContext()
+			var errs []error
+			for _, d := range types {
+				d := d
+
+				if types.IsArmDefinition(&d) {
+					// Skip ARM definitions, we don't need to create storage variants of those
+					continue
+				}
+
+				if types.IsEnumDefinition(&d) {
+					// Skip Enum definitions as we use the base type for storage
+					continue
+				}
+
+				def, err := visitor.VisitDefinition(d, vc)
+				if err != nil {
+					errs = append(errs, err)
+					continue
+				}
+
+				finalDefn := def.WithDescription(descriptionForStorageVariant(d))
+				storageTypes[finalDefn.Name()] = finalDefn
+
+			}
+
+			if len(errs) > 0 {
+				err := kerrors.NewAggregate(errs)
+				return nil, err
+			}
+
+			types.AddTypes(storageTypes)
+
+			return types, nil
+		},
+	}
+}
+
+// makeStorageTypesVisitor returns a TypeVisitor to do the creation of dedicated storage types
+func makeStorageTypesVisitor(types astmodel.Types) astmodel.TypeVisitor {
+	factory := &StorageTypeFactory{
+		types: types,
+	}
+
+	result := astmodel.MakeTypeVisitor()
+	result.VisitTypeName = factory.visitTypeName
+	result.VisitObjectType = factory.visitObjectType
+	result.VisitArmType = factory.visitArmType
+	//result.VisitResourceType = factory.visitResourceType
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types astmodel.Types
+}
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
```

Comment:
Is `actualDefinitionFound` false actually a fatal error here?

Could this be:
```
 	// Check for property specific handling
	if vc.property != nil {
               actualDefinition, actualDefinitionFound := factory.types[name]
               if !actualDefinitionFound {
                    // return err or panic
                }
		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
			// Property type refers to an enum, so we use the base type instead
			return et.BaseType(), nil
		}
	}
```

Reply:
It's not a fatal error because it might be referring to an imported type (not that I'm whether that's happening yet).

Reply:
It feels to me like it not being found deserves either a warning log or an error (I lean towards error, especially if you're not sure if it's happening at all).

If you make this return error and everything still runs, awesome. If not you can investigate the cause and if it's expected then maybe this is a warning or something.

Reply:
We definitely have imported types (aka a `TypeName` containing an `ExternalPackageReference` instead of a `LocalPackageReference`) that are legit, so emitting a warning would just be noise. (They're types imported from the `genruntime` package.)